test/distributed/org/apache/cassandra/distributed/test/sai/PartialUpdateHandlingTest.java

```diff
@@ -0,0 +1,541 @@
+import org.apache.cassandra.locator.InetAddressAndPort;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.apache.cassandra.distributed.api.ConsistencyLevel.ALL;
+import static org.apache.cassandra.distributed.api.ConsistencyLevel.QUORUM;
+import static org.apache.cassandra.distributed.api.Feature.GOSSIP;
+import static org.apache.cassandra.distributed.api.Feature.NETWORK;
+import static org.apache.cassandra.distributed.shared.AssertUtils.assertRows;
+import static org.apache.cassandra.index.sai.plan.Expression.IndexOperator.EQ;
+import static org.apache.cassandra.index.sai.plan.Expression.IndexOperator.RANGE;
+
+/**
+ * SAI queries, like all filtering queries, must correctly resolve divergent views of row data across replicas. In
+ * particular, cases where writes do not propagate to all replicas can, if not resolved correctly, lead to consistency
+ * violations, as rows that should not appear in our results do, while rows that should appear do not.
+ * <p> 
+ * The variables that affect the behavior of these queries are, at minimum, the following:
+ * <p> 
+ * 1.) The combination of write & read consistency levels used by the client. Reads at ONE/LOCAL_ONE trivially
+ *     avoid having to resolve data from diverging replicas.
+ * 2.) Interaction of existing data (if there is any), with partial updates and deletes. The same query that
+ *     erroneously returns stale matches with naïve resolution and partial updates on top of existing data might 
+ *     fail to return live matches with partial updates and no previous data.
+ * 3.) The number of query clauses and their targets. Clauses may target partition keys, clustering keys, static 
+ *     columns, and regular columns, and some combinations are more problematic than others.
+ * 4.) The repaired state of SSTables that participate in the query. A fully repaired on-disk set of SSTables cannot
+ *     produce erroneous results due to split rows (i.e. rows which contain partial updates of different columns
+ *     across different partitions).
+ * 5.) Whether data resides in SSTables or Memtables. The latter is implicitly unrepaired.
+ * 6.) Interaction w/ existing mechanisms on the distributed read path that deal with short reads, replica filtering
+ *     protection, etc.
+ * 7.) The relationship between columns selected and columns restricted by queries. (If coordinator filtering is
+ *     involved at the implementation level, retrieving enough information to do that filtering is important.)
+ * 8.) The timestamps of partial updates and deletes, especially for single-column queries that might produce
+ *     stale matches if not resolved correctly.
+ */
+@RunWith(Parameterized.class)
+public class PartialUpdateHandlingTest extends TestBaseImpl
+{
+    private static final String TEST_TABLE_NAME = "test_partial_updates";
+    private static final int PARTITIONS_PER_TEST = 10;
+    private static final int NODES = 3;
```

Comment:
Can we not reproduce all the problematic cases with just 2 nodes and save some resources?

Reply:
Two thoughts for me here:

1.) With the cluster being shared across all tests, one more node seems not to be that problematic.
2.) The QUORUM query cases require a 3rd node, right? Otherwise they're identical to the ALL query cases?

Reply:
I think all the scenarios of concern can be reproduced with `ONE`/`ALL`. I don't see how `QUORUM`/`QUORUM` can trigger something different. Half of the tested cases use `QUORUM` reads. If we consider `ONE`/`ALL` as a good enough model, we can drop the parameters using `QUORUM` reads and use a 2-node cluster. That would make the test more than twice as fast, saving CI resources.

Reply:
I'll reduce to ONE/ALL for now. I need to separately address the change in `IndexStatusManager` anyway...

See https://github.com/apache/cassandra/pull/2935#discussion_r1442402238